File: grid4_4.py
#ECSE420
#LAB2
#ANDONI ROMAN
#YORDAN NESHEV

#PYTHONIC DRUM 1.0


from mpi4py import MPI
import numpy as np
import sys

#GLOBALS
GRID_SIZE = 4
SIZE = GRID_SIZE * GRID_SIZE

#Node dictionary containing (i,j) of node as key and [t2,t1,t] as time values for each node
node_dict = {}
size = 0

#multiplicant constants
ETA = float(0.0002)
RHO = float(0.5)
G =	float(0.75)

#Dict that will get received by each node. Will contain all neigbouring values with their i and j
neigbour_nodes_and_their_values = {}
dots_per_process = 0

#Fills up all information necessary for node transactions
#Each node will contain its position i j and an array with 4 variables
#that contain values from neighbours

#In case of 4x4 we will have 1 dictionary assigned since one dot is assigned to a process
#Otehrwise, we will have a many variables due to the fact that we assign a cluster of dots
#to a process. Dictionary has (i,j) position as key and [n2,n1,n0] as value
#indicating previous-previous value, previous value, current value
#STRIKE value is set to be n2 at beginning at node N/2,N/2 --> [0,0,1]
def fill_node_dict(dot_amount_per_process, rank):

    # case N=4
    if dots_per_process == 1:
        for z in range (0, dots_per_process):
            rank_counter = -1
            for j in range (0,GRID_SIZE):
                for i in range(0, GRID_SIZE):
                    rank_counter += 1
                    if rank_counter == rank:
                        if (i == GRID_SIZE/2) and (j == GRID_SIZE/2):
                            node_dict[(i,j)] = [0.0, 0.0, 0.0]
                        else:
                            node_dict[(i,j)] = [0.0, 0.0, 0.0]
    #case N=512
    else:
        offset = GRID_SIZE/size
        beginning = (offset * rank)
        end = beginning + offset
        for j in range(beginning, end):
            for i in range(0, GRID_SIZE):
                if (i == GRID_SIZE / 2) and (j == GRID_SIZE / 2):
                    node_dict[(i, j)] = [0.0, 0.0, 0.0]
                else:
                    node_dict[(i, j)] = [0.0, 0.0, 0.0]

#Returns the number of the node in 2D array starting from 0 to N-1*N-1
def get_node_num(coord_tuple):
    #eliminate the boundary conditions first
    num = 0

    if(coord_tuple[0] < 0 or coord_tuple[1]  < 0):
        num = -1
    elif(coord_tuple[0] >= GRID_SIZE or coord_tuple[0] >= GRID_SIZE):
        num = -1
    else:
        rank_counter = -1
        for j in range(0, GRID_SIZE):
            for i in range(0, GRID_SIZE):
                rank_counter += 1
                if(coord_tuple[0]==i and coord_tuple[1]==j):
                    num = rank_counter
    return num


#Returns the ranks of the process that contains the specified neighbor coordinate
def get_neighbor_rank(coord_tuple, dots_per_process):
    #eliminate the boundary conditions first
    rank = 0

    if(coord_tuple[0] < 0 or coord_tuple[1]  < 0):
        rank = -1
    elif(coord_tuple[0] >= GRID_SIZE or coord_tuple[0] >= GRID_SIZE):
        rank = -1
    else:
        rank_counter = -1
        for j in range(0, GRID_SIZE):
            for i in range(0, GRID_SIZE):
                rank_counter += 1
                if(coord_tuple[0]==i and coord_tuple[1]==j):
                    rank = rank_counter
                    rank = int(rank/dots_per_process)
    return rank

#Returns neigbour coordinate based on current coordinate
# k ==>  0:left 1:right 2: up 3: down
def get_neighbor_coord_based_on_current_coordinate(i,j,k):

    neighbor_coordinate = (-1,-1)

    if k == 0:
        neighbor_coordinate = (i-1,j)
    elif k == 1:
        neighbor_coordinate = (i+1,j)
    elif k == 2:
        neighbor_coordinate = (i, j+1)
    elif k == 3:
        neighbor_coordinate = (i, j-1)

    return neighbor_coordinate

#Based on conditions verifies if the node provided the node number needs update
#Returns true if update needed
def check_if_needs_update(condition, node_number):

    #CORNER CASE
    if (node_number == 0 or node_number == GRID_SIZE - 1 or node_number == GRID_SIZE * (GRID_SIZE - 1) or node_number == ( GRID_SIZE * GRID_SIZE - 1)):
        if(condition != "CORNER"):
            return False
    elif (node_number % GRID_SIZE == 0 or node_number % GRID_SIZE == GRID_SIZE - 1 or  node_number < GRID_SIZE or  node_number > GRID_SIZE * (GRID_SIZE - 1)):
        if(condition != "EDGE"):
            return False
    else:
        if(condition != "CENTER"):
            return False
    return True



def exchange_data_with_neighbors(rank, buffer, operation=None, condition=None):

    counter = 0
    key_list = node_dict.keys()

    for i in range (0, GRID_SIZE):
        for j in range (0, GRID_SIZE):

            if ((i,j) in key_list):
                kcount = 0
                #check for neighbors 0:left 1:right 2: up 3: down
                for k in range (0,4):

                    #getting neighbors coordinate
                    neighbour_coord = get_neighbor_coord_based_on_current_coordinate(i,j,k)

                    #SENDING TO NEIGHBORS
                    if operation == 'send':
                        #Getting neighbors rank
                        rank_of_neighbor_to_send_to = get_neighbor_rank(neighbour_coord,dots_per_process)

                        #if not outside the grid and not same rank as current process
                        #then send
                        #Applies to one process per dot or many dots per process
                        if ((rank_of_neighbor_to_send_to != -1) and (rank_of_neighbor_to_send_to != rank)):

                            neighbour_node_num = get_node_num(neighbour_coord)

                            if (check_if_needs_update(condition, neighbour_node_num) == False):
                                continue
                            #otherwise, the neighbour belongs to other process. Need to send.
                            #get our own value and then send to the neighbor
                            #send the previous value (t-1) value
                            my_value_to_send = node_dict[(i,j)][1]
                            my_tag = get_node_num((i,j))
                            #MPI SEND
                            comm.send(my_value_to_send, dest=rank_of_neighbor_to_send_to, tag=my_tag);
                            counter += 1

                    # RECEIVING FROM NEIGHBORS
                    elif operation == 'recv':
                        rank_of_neighbor_to_receive_from = get_neighbor_rank(neighbour_coord,dots_per_process)
                        # if not outside the grid and not same rank as current process
                        # then receive
                        # Applies to one process per dot or many dots per process
                        if ((rank_of_neighbor_to_receive_from != -1) and (rank_of_neighbor_to_receive_from != rank)):

                            my_node_number = get_node_num((i,j))
                            if (check_if_needs_update(condition, my_node_number) == False):
                                continue
                            #otherwise we NEED update so need to receive from process
                            neighbours_tag = get_node_num((neighbour_coord))
                            value_to_receive = comm.recv(source =rank_of_neighbor_to_receive_from, tag = neighbours_tag)
                            #updating neighbour dict with t-1 value of neighbour at (i,j)
                            neigbour_dict_keys = neigbour_nodes_and_their_values.keys()

                            #will create a new entry at first run if neighbour coordinate and value array is not
                            #in neighbour dictionary yet
                            if(not(neighbour_coord in neigbour_dict_keys)):
                                neigbour_nodes_and_their_values[neighbour_coord] = [0.0,0.0,0.0]

                            neigbour_nodes_and_their_values[neighbour_coord][1]=value_to_receive

                        #Means those nodes belong to US (1 process with many dots). Simply distribute between neighbours
                        #without MPI sending.
                        elif rank_of_neighbor_to_receive_from == rank:
                            value_to_receive = node_dict[neighbour_coord][1]








#simulation iteration method. Will update nodes and chose nodes based on condition provided
#default is non but specified from calls from main. Can be: EDGE, CENTER, CORNER
def simulate_iteration(rank,condition = None):

    #All received values from the neighbours will be stored here
    #will always have 4 structures with  values in it:
        #UPPER NEIGBOURS i,j as key and their array with PREV-PREV, PREV, CURRENT values
        #LOWER NEIGBOURS i,j as key and their array with PREV-PREV, PREV, CURRENT values
        #LEFT NEIGHBOURS i,j as key and their array with PREV-PREV, PREV, CURRENT values
        #RIGHT NEIGHBOURS i,j as key and their array with PREV-PREV, PREV, CURRENT values
    #depending on condition, there will be special cases (ex: corner cases)

    exchange_data_with_neighbors(rank, {}, operation= 'send', condition = condition)

    exchange_data_with_neighbors(rank, {},  operation = 'recv', condition = condition)

    #Now update depending on condition
    #get now list of all keys of nodes  that belong to process
    #Based on this key list, each process will update only those nodes that belong to it
    #So we check the key first, if its in processes pool of keys.
    #If the key is in the pool, check the received neighbor values.
    #Take the value (t2, t1 or t0) basing on the case (corner, middle or edges)
    # ex neighbor_nodes_and_their_values[(1,0)][2] means we get the most recent value (position 2 in value array)
    #  of  neighbor that is situated at 1,0

    key_list = node_dict.keys()

    neigbour_key_list = neigbour_nodes_and_their_values.keys()

    if(condition == "CORNER"):
        #case upper left corner. depends on previous value of node i=1,j=0
        if (0,0) in key_list:
            if ((1,0) in neigbour_key_list):
                update_value = G * neigbour_nodes_and_their_values[(1,0)][2]
                node_dict[(0,0)][2] = update_value

        # case upper right corner. depends on previous value of node i=N-2,j=0
        elif (GRID_SIZE-1, 0) in key_list:
            if ((GRID_SIZE-2, 0) in neigbour_key_list):
                update_value = G * neigbour_nodes_and_their_values[(GRID_SIZE-2, 0)][2]
                node_dict[(GRID_SIZE-1, 0)][2] = update_value

        # case lower  left corner. depends on previous value of node i=0,j=N-2
        elif (0, GRID_SIZE-1) in key_list:
            if ((0, GRID_SIZE-2) in neigbour_key_list):
                update_value = G * neigbour_nodes_and_their_values[(0, GRID_SIZE-2)][2]
                node_dict[(0,GRID_SIZE-1)][2] = update_value

        # case lower right corner. depends on previous value of node i=N-1,j=N-2
        elif (GRID_SIZE-1, GRID_SIZE-1) in key_list:
            if ((GRID_SIZE-1, GRID_SIZE-2) in neigbour_key_list):
                update_value = G * neigbour_nodes_and_their_values[(GRID_SIZE-1, GRID_SIZE-2)][2]
                node_dict[(GRID_SIZE-1,GRID_SIZE-1)][2] = update_value


    elif(condition == "EDGE"):
        #left most edge case --> get all nodes with i at 0 and j starting from 1 till N-2
        for j in range(1,GRID_SIZE-1):
            if (0,j) in key_list:
                if(1,j) in neigbour_key_list:
                    update_value = G * neigbour_nodes_and_their_values[(1,j)][2]
                    node_dict[(0,j)][2] = update_value

        #right  most edge case --> get all nodes with i at N-1 j starting from 1 till N-2
        for j in range(1,GRID_SIZE-1):
            if (GRID_SIZE-1,j) in key_list:
                if (GRID_SIZE-2,j) in neigbour_key_list:
                    update_value = G * neigbour_nodes_and_their_values[(GRID_SIZE-2,j)][2]
                    node_dict[(GRID_SIZE-1,j)][2] = update_value

        #right top edge case --> get all nodes with j at 0 and x starting from 1 till N-2
        for i in range(1,GRID_SIZE-1):
            if (i,0) in key_list:
                if (i,1) in neigbour_key_list:
                    update_value = G * neigbour_nodes_and_their_values[(i,1)][2]
                    node_dict[(i,0)][2] = update_value

        #bottom top edge case --> get all nodes with j at 0 and x starting from 1 till N-2
        for i in range(1,GRID_SIZE-1):
            if (i,GRID_SIZE-1) in key_list:
                if (i,GRID_SIZE-2) in neigbour_key_list:
                    update_value = G * neigbour_nodes_and_their_values[(i,GRID_SIZE-2)][2]
                    node_dict[(i,GRID_SIZE-1)][2] = update_value

    elif(condition == "CENTER"):
        #CENTER CASE
        #All i's are situated between 1 and N-2
        #All j's are situated between 1 and N-2
        #Get all 4 neigbours values and then update the value using the formula
        for i in range(1,GRID_SIZE-1):
            for j in range(1,GRID_SIZE-1):
                if(i,j) in key_list:

                    left_neigbour_prev_val = 0
                    right_neighbour_prev_val = 0
                    lower_neighbour_prev_val = 0
                    upper_neighbour_prev_val = 0
                    #now treat every neighbour separately
                    if(i-1,j) in neigbour_key_list:
                        left_neigbour_prev_val = neigbour_nodes_and_their_values[(i-1,j)][1]
                    if(i+1,j) in neigbour_key_list:
                        right_neighbour_prev_val = neigbour_nodes_and_their_values[(i+1,j)][1]
                    if(i, j-1) in neigbour_key_list:
                        lower_neighbour_prev_val = neigbour_nodes_and_their_values[(i, j-1)][1]
                    if(i, j+1) in neigbour_key_list:
                        upper_neighbour_prev_val = neigbour_nodes_and_their_values[(i, j+1)][1]

                    glorious_center_value = (RHO * (left_neigbour_prev_val + right_neighbour_prev_val + \
                            upper_neighbour_prev_val + lower_neighbour_prev_val - (4*node_dict[(i,j)][1])) + \
                            2*(node_dict[(i,j)][1]) - (1 - ETA)*(node_dict[(i,j)][0]))/(1-ETA)
                    #update to most recent value
                    node_dict[(i,j)][2] = glorious_center_value



if __name__ == "__main__":

    iterations = int(sys.argv[1])

    #MPI inits
    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()
    size = comm.Get_size()
    status = MPI.Status()

    dots_per_process = (GRID_SIZE*GRID_SIZE / size)

    #setup dict for each node. The dict will contain (i,j) tuple as key and array [n2,n1,n0]
    #for t-2,t-1,t values
    fill_node_dict(dots_per_process,rank)


    # For each of iteration, do 3 updates. First center, then edges, then corners.
    # Then print the value at N/2 N/2 where strike was applied
    for iteration in range(0,iterations + 1):

        simulate_iteration(rank,condition = "CENTER")

        simulate_iteration(rank,condition = "EDGE")

        simulate_iteration(rank,condition = "CORNER")

        print "I'm rank " + str(rank) + " and my grid dict is" + str(node_dict) + '\n\n'

        #STRIKE PRINT
        key_list = node_dict.keys()
        if((GRID_SIZE/2,GRID_SIZE/2) in key_list):
            #strike value is the value we get from node dict. The value is the array with n2,n1,n0. We display n0
            #the latest (current) strike value
            strike_value = node_dict[(GRID_SIZE/2,GRID_SIZE/2)][2]
            print('Strike value at iteration ' + str(iteration) + ' is ' + str(strike_value))

        iterations -= 1

        #at each iteration also shift previous values.
    MPI.Finalize()

File: grid4_4.c
#include <stdio.h>
#include <stdlib.h>
#include "mpi.h"
#include <time.h>
#include <unistd.h>
#include <stdint.h>
#include <sys/time.h>
#include <math.h>

#define SEND_TAG 0
#define ANSWER_TAG 1
#define COMMUNICATE_TAG 2
#define GRID_SIZE 4
#define UPDATE_INTERIOR 0
#define UPDATE_BOUNDARY 1
#define UPDATE_CORNER 2
#define P 0.5
#define N 0.0002
#define G 0.75

typedef struct {
	int rank;
    float value;
	float prev_value;
	float prev_prev_value;
} gridNode;

//Current Grid
gridNode* grid[GRID_SIZE][GRID_SIZE];
//[0]=top [1]=right [2]=bottom [3]=left
float neighbor_nodes_values[4];
int numberOfSlaves;
//If printComment==1, then print some debug comments
int printComments = -1;


// Each node in the 4x4 grid is an independent process
void setupGrid(int rank){
	//Allocate memory to the global array of gridNode pointers (16 pointers)
	//grid = malloc(sizeof(gridNode *) * GRID_SIZE * GRID_SIZE);
	int row;
	int column;
	for (row=0;row<GRID_SIZE;row++){
		for(column=0;column<GRID_SIZE;column++){
			// Allocate memory to each gridNode pointer that will contain the info relative to a node
			grid[row][column] = malloc(sizeof(gridNode));
			
			//Fill gridNodes
			if(row == GRID_SIZE/2 && column == GRID_SIZE/2){
				//Center node (Set prev_value to 1.0 to simulate a hit on the drum)
				grid[row][column]->rank = rank;
				grid[row][column]->value = 1.0;
				grid[row][column]->prev_value = 0.0;
				grid[row][column]->prev_prev_value = 0.0;
			}else{
				grid[row][column]->rank = rank;
				grid[row][column]->value = 0.0;
				grid[row][column]->prev_value = 0.0;
				grid[row][column]->prev_prev_value = 0.0;
			}
		}
	}
}

int getRowFromRank(int rank) {
	int row = floor(rank / GRID_SIZE);
	return row;
}

int getColumnFromRank(int rank) {
	int column = rank % GRID_SIZE;
	return column;
}

int getRankFromRowColumn(int row, int column){
	if(row<0 || row>GRID_SIZE-1 || column<0 || column>GRID_SIZE-1){
		//Wrong coordinates
		return -1;
	}else{
		return row*GRID_SIZE+column;
	}
}

//Return 1 if coordinates point to corner node, 0 otherwise
int isCornerNode(int rank){
	int row = getRowFromRank(rank);
	int column = getColumnFromRank(rank);
	if((row==0 && column==0) || (row==0 && column==GRID_SIZE-1) || (row==GRID_SIZE-1 && column==0) || (row==GRID_SIZE-1 && column==GRID_SIZE-1)){
		return 1;
	}else{
		return 0;
	}
}

//Return 1 if coordinates point to boundary node, 0 otherwise
int isBoundaryNode(int rank){
	int row = getRowFromRank(rank);
	int column = getColumnFromRank(rank);
	//Boundary, but not corner
	if((row==0 || row==GRID_SIZE-1 || column==0 || column==GRID_SIZE-1) && (isCornerNode(rank))!=1){
		return 1;
	}else{
		return 0;
	}
}

//Return 1 if coordinates point to interior node, 0 otherwise
int isInteriorNode(int rank){
	int row = getRowFromRank(rank);
	int column = getColumnFromRank(rank);
	if((row>0 && row<GRID_SIZE-1) && (column>0 && column<GRID_SIZE-1)){
		return 1;
	}else{
		return 0;
	}
}

//Position: [0]=top [1]=right [2]=bottom [3]=left
int getRankNeighborNode(int myRow, int myColumn, int position){
	if(position==0){
		//Top
		return getRankFromRowColumn(myRow-1, myColumn);
	}else if(position==1){
		//Right
		return getRankFromRowColumn(myRow, myColumn+1);
	}else if(position==2){
		//Bottom
		return getRankFromRowColumn(myRow+1, myColumn);
	}else if(position==3){
		//Left
		return getRankFromRowColumn(myRow, myColumn-1);
	}else {
		//Wrong position value, no valid neighbor
		return -1;
	}
}

void performDataExchangesForInteriorHelper(int rank, int operation){
	//Get previous values from all boundary nodes
	//getRankNeighborNode [0]=top [1]=right [2]=bottom [3]=left
	int k;
	int row = getRowFromRank(rank);
	int column = getColumnFromRank(rank);
	if(operation==SEND_TAG){
		if(isBoundaryNode(rank)){
			for(k=0;k<4;k++){
				int neighborRankToSendTo = getRankNeighborNode(row, column, k);
				if(isInteriorNode(neighborRankToSendTo)==1){
					//Boundary node sends its previous value to a neighbor interior node
					MPI_Send(&(grid[row][column]->prev_value), 1, MPI_FLOAT, neighborRankToSendTo, COMMUNICATE_TAG, MPI_COMM_WORLD);
					//printf("I am rank %d and sent value %f to neigbor rank %d.\n", rank, grid[row][column]->prev_value, neighborRankToSendTo);
				}
			}
		}else if(isInteriorNode(rank)==1){
			for(k=0;k<4;k++){
				int neighborRankToSendTo = getRankNeighborNode(row, column, k);
				if(isInteriorNode(neighborRankToSendTo)==1){
					//Interior node send its previous value to another interior node
					MPI_Send(&(grid[row][column]->prev_value), 1, MPI_FLOAT, neighborRankToSendTo, COMMUNICATE_TAG, MPI_COMM_WORLD);
					//printf("I am rank %d and sent value %f to neigbor rank %d.\n", rank, grid[row][column]->prev_value, neighborRankToSendTo);
				}
			}
		}
	}else if(operation==ANSWER_TAG && isInteriorNode(rank)==1){
		//Interior nodes receive values from all of their neighbors
		for(k=0;k<4;k++){
			int neighborRankToReceiveFrom = getRankNeighborNode(row, column, k);
			MPI_Recv(&(neighbor_nodes_values[k]), 1, MPI_FLOAT, neighborRankToReceiveFrom, COMMUNICATE_TAG, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
			//printf("I am rank %d and received value %f from neigbor rank %d.\n", rank, neighbor_nodes_values[k], neighborRankToReceiveFrom);
		}
	}
}

void performDataExchangesForInterior(int rank){
	performDataExchangesForInteriorHelper(rank, SEND_TAG);
	MPI_Barrier(MPI_COMM_WORLD);
	performDataExchangesForInteriorHelper(rank, ANSWER_TAG);
}

void performDataExchangesForBoundaryHelper(int rank, int operation){
	//Send to boundary nodes needed values from neighbor interior nodes
	//getRankNeighborNode [0]=top [1]=right [2]=bottom [3]=left
	int k;
	int row = getRowFromRank(rank);
	int column = getColumnFromRank(rank);
	if(operation==SEND_TAG && isInteriorNode(rank)==1){
		for(k=0;k<4;k++){
			int neighborRankToSendTo = getRankNeighborNode(row, column, k);
			if(isBoundaryNode(neighborRankToSendTo)==1){
				//Interior node sends its current value to a neighbor boundary node
				MPI_Send(&(grid[row][column]->value), 1, MPI_FLOAT, neighborRankToSendTo, COMMUNICATE_TAG, MPI_COMM_WORLD);
				//printf("I am rank %d and sent value %f to neigbor rank %d.\n", rank, grid[row][column]->value, neighborRankToSendTo);
			}
		}
	}else if(operation==ANSWER_TAG && isBoundaryNode(rank)==1){
		//Boundary nodes receive values from neighbor interior nodes
		for(k=0;k<4;k++){
			int neighborRankToReceiveFrom = getRankNeighborNode(row, column, k);
			if(isInteriorNode(neighborRankToReceiveFrom)==1){
				MPI_Recv(&(neighbor_nodes_values[k]), 1, MPI_FLOAT, neighborRankToReceiveFrom, COMMUNICATE_TAG, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
				//printf("I am rank %d and received value %f from neigbor rank %d.\n", rank, neighbor_nodes_values[k], neighborRankToReceiveFrom);
			}
		}
	}
}

void performDataExchangesForBoundary(int rank){
	performDataExchangesForBoundaryHelper(rank, SEND_TAG);
	MPI_Barrier(MPI_COMM_WORLD);
	performDataExchangesForBoundaryHelper(rank, ANSWER_TAG);
}

void performDataExchangesForCornersHelper(int rank, int operation){
	//Send to boundary nodes needed values from neighbor interior nodes
	//getRankNeighborNode [0]=top [1]=right [2]=bottom [3]=left
	int k;
	int row = getRowFromRank(rank);
	int column = getColumnFromRank(rank);
	if(operation==SEND_TAG && isBoundaryNode(rank)==1){
		//Specific boundary nodes send their current value to neighbor corner nodes
		int neighborRankToSendTo=-1;
		if(row==1 && column==0){
			//Send my current value to [0,0] corner
			neighborRankToSendTo = getRankFromRowColumn(0,0);
		}else if(row==GRID_SIZE-2 && column==0){
			//Send my current value to [GRID_SIZE-1,0] corner
			neighborRankToSendTo = getRankFromRowColumn(GRID_SIZE-1,0);
		}else if(row==0 && column==GRID_SIZE-2){
			//Send my current value to [0,GRID_SIZE-1] corner
			neighborRankToSendTo = getRankFromRowColumn(0,GRID_SIZE-1);
		}else if(row==GRID_SIZE-1 && column==GRID_SIZE-2){
			//Send my current value to [GRID_SIZE-1,GRID_SIZE-1] corner
			neighborRankToSendTo = getRankFromRowColumn(GRID_SIZE-1,GRID_SIZE-1);
		}
		if(neighborRankToSendTo!=-1){
			MPI_Send(&(grid[row][column]->value), 1, MPI_FLOAT, neighborRankToSendTo, COMMUNICATE_TAG, MPI_COMM_WORLD);
			//printf("I am rank %d and sent value %f to neigbor rank %d.\n", rank, grid[row][column]->value, neighborRankToSendTo);
		}
	}else if(operation==ANSWER_TAG && isCornerNode(rank)==1){
		//Corner nodes receive values from specific neighbor boundary nodes
		int neighborRankToReceiveFrom=-1;
		if(row==0 && column==0){
			//Receive current value from bottom [1,0] boundary node
			neighborRankToReceiveFrom = getRankFromRowColumn(1,0);
			k=2;
		}else if(row==GRID_SIZE-1 && column==0){
			//Receive current value from top [GRID_SIZE-2,0] boundary node
			neighborRankToReceiveFrom = getRankFromRowColumn(GRID_SIZE-2,0);
			k=0;
		}else if(row==0 && column==GRID_SIZE-1){
			//Receive current value from left [0,GRID_SIZE-2] boundary node
			neighborRankToReceiveFrom = getRankFromRowColumn(0,GRID_SIZE-2);
			k=3;
		}else if(row==GRID_SIZE-1 && column==GRID_SIZE-1){
			//Receive current value from left [GRID_SIZE-1,GRID_SIZE-2] boundary node
			neighborRankToReceiveFrom = getRankFromRowColumn(GRID_SIZE-1,GRID_SIZE-2);
			k=3;
		}
		if(neighborRankToReceiveFrom!=-1){
			MPI_Recv(&(neighbor_nodes_values[k]), 1, MPI_FLOAT, neighborRankToReceiveFrom, COMMUNICATE_TAG, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
			//printf("I am rank %d and received value %f from neigbor rank %d.\n", rank, neighbor_nodes_values[k], neighborRankToReceiveFrom);
		}
	}
}

void performDataExchangesForCorners(int rank){
	performDataExchangesForCornersHelper(rank, SEND_TAG);
	MPI_Barrier(MPI_COMM_WORLD);
	performDataExchangesForCornersHelper(rank, ANSWER_TAG);
}

void setNewValue(float newValue, int row, int column){
	grid[row][column]->prev_prev_value = grid[row][column]->prev_value;
	grid[row][column]->prev_value = grid[row][column]->value;
	grid[row][column]->value = newValue;
}

void perform_iterationHelper(int rank, int operation){
	int row = getRowFromRank(rank);
	int column = getColumnFromRank(rank);
	
	//Update Interior Nodes
	if(operation == UPDATE_INTERIOR){
		performDataExchangesForInterior(rank);
		if(isInteriorNode(rank)==1){
			float newValue = (P*(neighbor_nodes_values[0]+neighbor_nodes_values[1]+neighbor_nodes_values[2]+neighbor_nodes_values[3]-4*grid[row][column]->prev_value)+2*grid[row][column]->prev_value-(1-N)*grid[row][column]->prev_prev_value)/(1+N);
			setNewValue(newValue, row, column);
		}
	}
	
	//Update Boundary Nodes
	if(operation == UPDATE_BOUNDARY){
		performDataExchangesForBoundary(rank);
		if(isBoundaryNode(rank)==1){
			//[0]=top [1]=right [2]=bottom [3]=left
			float newValue;
			if(row==0){
				newValue = G*neighbor_nodes_values[2];
			}else if(row==GRID_SIZE-1){
				newValue = G*neighbor_nodes_values[0];
			}else if(column==0){
				newValue = G*neighbor_nodes_values[1];
			}else if(column==GRID_SIZE-1){
				newValue = G*neighbor_nodes_values[3];
			}
			setNewValue(newValue, row, column);
		}
	}
	
	//Update Corner Nodes
	if(operation == UPDATE_CORNER){
		performDataExchangesForCorners(rank);
		if(isCornerNode(rank)==1){
			//[0]=top [1]=right [2]=bottom [3]=left
			float newValue;
			if(row==0 && column==0){
				newValue = G*neighbor_nodes_values[2];
			}else if(row==GRID_SIZE-1 && column==0){
				newValue = G*neighbor_nodes_values[0];
			}else if(row==0 && column==GRID_SIZE-1){
				newValue = G*neighbor_nodes_values[3];
			}else if(row==GRID_SIZE-1 && column==GRID_SIZE-1){
				newValue = G*neighbor_nodes_values[3];
			}
			setNewValue(newValue, row, column);
		}
	}

	MPI_Barrier(MPI_COMM_WORLD);
}

void shiftValues(int rank){
	int row = getRowFromRank(rank);
	int column = getColumnFromRank(rank);
	grid[row][column]->prev_prev_value = grid[row][column]->prev_value;
	grid[row][column]->prev_value = grid[row][column]->value;
	MPI_Barrier(MPI_COMM_WORLD);
}

void perform_iteration(int rank){
	shiftValues(rank);
	perform_iterationHelper(rank, UPDATE_INTERIOR);
	perform_iterationHelper(rank, UPDATE_BOUNDARY);
	perform_iterationHelper(rank, UPDATE_CORNER);
}

void printMiddleValue(int iteration, int rank){
	//Print value at [GRID_SIZE/2,GRID_SIZE/2]
	if(rank==getRankFromRowColumn(GRID_SIZE/2, GRID_SIZE/2)){
		printf("Iteration %d: %f\n", iteration, grid[GRID_SIZE/2][GRID_SIZE/2]->value);
	}
}

//If includeRank=1, also include rank values
//If includePrevValues=1, also include previous values
void printGrid(int rank, int includeRank, int includePrevValues){
	MPI_Barrier(MPI_COMM_WORLD);
	//Send grid Elements to master
	int row = getRowFromRank(rank);
	int column = getColumnFromRank(rank);
	char textOutputBuffer[256];
	
	if(includeRank==1 && includePrevValues==1){
		//Print (x,y) coordinates, rank, present value and previous values
		sprintf( textOutputBuffer, "(%d,%d): [%d,%f,%f,%f]", row, column, grid[row][column]->rank, grid[row][column]->value, grid[row][column]->prev_value, grid[row][column]->prev_prev_value);
	}else if(includeRank==1){
		//Print (x,y) coordinates, rank and present value
		sprintf( textOutputBuffer, "(%d,%d): [%d,%f]", row, column, grid[row][column]->rank, grid[row][column]->value);
	}else if(includePrevValues==1){
		//Print (x,y) coordinates, present value and previous values
		sprintf( textOutputBuffer, "(%d,%d): [%f,%f,%f]", row, column, grid[row][column]->value, grid[row][column]->prev_value, grid[row][column]->prev_prev_value);
	}else {
		//Print only (x,y) coordinates and present value
		sprintf( textOutputBuffer, "(%d,%d): %f", row, column, grid[row][column]->value);
	}
	MPI_Send(&textOutputBuffer, 256, MPI_CHAR, 0, ANSWER_TAG, MPI_COMM_WORLD);
	
	if(rank == 0){
		//Print the grid from the master by receiving results from all processes
		printf("\n\n");
		int i;
		for (i = 0; i <= numberOfSlaves; i++) {	
			//Receive one message from the master and one message from each slave
			char textOutputBuffer[256];
			MPI_Recv(&textOutputBuffer, 256, MPI_CHAR, i, ANSWER_TAG, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
			printf("%s   ", textOutputBuffer);
			if(getColumnFromRank(i) == GRID_SIZE-1){
				//New line
				printf("\n");
			}
		}
		printf("\n\n");
	}
	MPI_Barrier(MPI_COMM_WORLD);
}

void printProcessInfo(int rank){
	MPI_Barrier(MPI_COMM_WORLD);
	//Send process info to master
	int row = getRowFromRank(rank);
	int column = getColumnFromRank(rank);
	char textOutputBuffer[256];
	sprintf( textOutputBuffer, "I am process rank %d with:  value = %f    prev_value = %f    prev_prev_value = %f\n", grid[row][column]->rank, grid[row][column]->value, grid[row][column]->prev_value, grid[row][column]->prev_prev_value);
	MPI_Send(&textOutputBuffer, 256, MPI_CHAR, 0, ANSWER_TAG, MPI_COMM_WORLD);
	
	if(rank == 0){
		//Print the process info from the master by receiving results from all processes
		printf("\n\n");
		int i;
		for (i = 0; i <= numberOfSlaves; i++) {	
			//Receive one message from the master and one message from each slave
			char textOutputBuffer[256];
			MPI_Recv(&textOutputBuffer, 256, MPI_CHAR, i, ANSWER_TAG, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
			printf("%s\n", textOutputBuffer);
		}  
		printf("\n\n");
	}
	MPI_Barrier(MPI_COMM_WORLD);
}



int main(int argc, char** argv) {
    // Initialize the MPI environment
    MPI_Init(&argc, &argv);

    // Get the number of processes
    int numberOfProcesses;
    MPI_Comm_size(MPI_COMM_WORLD, &numberOfProcesses);

	//Test if at least one process provided
	if (numberOfProcesses < 1 )
	{
	  printf("ERROR! Need at least one process!\n");
	  MPI_Abort(MPI_COMM_WORLD, 1);
	  exit(1);
	}
	
	numberOfSlaves = numberOfProcesses - 1;

    // Get the rank of the process
	int rank;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
	
	setupGrid(rank);
	
	//If includeRank=1, also include rank values
	//If includePrevValues=1, also include previous values
	printGrid(rank,0,0);
	//printProcessInfo(rank);
	
	int iterations = atoi(argv[1]);
	int i;
	for(i=1;i<=iterations;i++){
		perform_iteration(rank);
		printGrid(rank,0,0);
		printMiddleValue(i, rank);
		//printProcessInfo(rank);
	}

    // Finalize the MPI environment.
    MPI_Finalize();
}